### PR TITLE
[Wasm GC] Update nulls to allow finding better LUBs

### PR DESCRIPTION
**Pull request by kripken**
This replaces most of #4302 : it does the null updating part, but it does not
add anything for fallthrough values, and the API is rewritten to be a lot clearer.
There is now `getBestPossible()` which checks what the best LUB is, and
`updateNulls()` which is explicit about when nulls are updated.

That is the diff against the other PR. The following is a proper description
from scratch that is suitable for reading this without that PR, and that will
be used in the final squashed commit.

It is common in GC code to have stuff like this:

```js
x = null;
..
x = Data();
```

Nulls in wasm have a type, and if that initial null has say anyref then
before this PR we would keep the type of x as anyref. However,
while nulls have types, all null values are identical, and so we can
in fact change x's type to a nullable reference of Data, by also
changing the null's type to something more specific.

LUBFinder now has an API that can return the best possible LUB
so far, and that can be told to update nulls if we decide that the
new LUB is worth using. This updates the passes using LUBFinder
to use the new API. Note how TypeRefining becomes simpler
because the special logic it had in a subclass of LUBFinder is now
part of the main class (it used to remember if there was a null
default; LUBFinder now handles both a null default as well as
other nulls).

This requires some changes to existing tests to avoid them from
optimizing using nulls in ways that ends up not testing the
original intent. Specifically the `dae-gc-refine-params.wast` now
has calls to get a null of a type, instead of just having a `ref.null`
of that type (which could be optimized now). And
`dae-gc-refine-return` wraps nulls in blocks.

### CODE REVIEW

**Review by tlively — COMMENTED** [review]
Could we put the method implementations in a C++ file instead of lubs.h?

**Comment by kripken** [comment]
> Could we put the method implementations in a C++ file instead of lubs.h?

We could, but aside from `updateNulls` and `combine` I think we need this all to inline. That is, the note/get/etc. methods need to be fast as we'll be calling them a lot. Given that, I'm not sure it's worth a cpp file just for a small subset of these methods?

File: src/ir/lubs.h [hunk 1]
```diff
@@ -22,31 +22,113 @@
...
+  // do not contribute type information - we do not try to find a lub based on
+  // them (rather we update them to the LUB).
   bool noted() { return lub != Type::unreachable; }
 
-  // Returns the lub that we found.
-  Type get() { return lub; }
+  // Returns the best possible lub. This ignores updatable nulls for the reasons
+  // mentioned above, since they will not limit us, aside from making the type
+  // nullable if nulls exist. This does not update the nulls.
+  Type getBestPossible() {
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/ir/lubs.h:67]: This goes back to the issue we ran into in the past PR. As you correctly noted there, the API was misleading. This class used to be read-only, but the new work makes it also alter things, and so I think it's very important to clarify when it does any updating. `getBestPossible()` does that, I think, by indicating it is just the *possible* LUB, but no commitment has been made yet as to actually using that value and making any updates anywhere. And `updateNulls` is likewise explicit about being a call that actually makes changes.

File: src/ir/lubs.h [hunk 2]
```diff
@@ -22,31 +22,113 @@
...
+
+    return changed;
+  }
+
+  // Force the lub to a particular type. This resets the entire state of the
+  // LUBFinder to just contain that type.
+  void set(Type type) {
+    lub = type;
+    nulls.clear();
   }
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/ir/lubs.h]: Yes, perhaps this is kind of strange. I refactored to remove this method now.

(We can't use note at the places that called this as that would not alter the nulls, but we can just replace the entire LUBFinder there, which does seem clearer.)

File: test/lit/passes/dae-gc-refine-params.wast [hunk 3]
```diff
@@ -81,41 +83,41 @@
...
+ ;; NOMNL-NEXT:   (call $get_null_{i32_i64})
  ;; NOMNL-NEXT:  )
  ;; NOMNL-NEXT: )
  (func $call-various-params-yes
   ;; The first argument gets {i32} and {i32}; the second {i32} and {i32_i64};
   ;; both of those pairs can be optimized to {i32}.
   ;; There is also an i32 in the middle, which should not confuse us.
   (call $various-params-yes
-   (ref.null ${i32})
+   (call $get_null_{i32})
```
[the remark this answers is not in the document]
Comment (kripken) [reply test/lit/passes/dae-gc-refine-params.wast:113]: Hmm, `get_{i32}()` suggests to me a function that returns `(ref ${i13})`, but here we want `(ref null ${i32})`, so the `null` in the name clarifies things?

File: test/lit/passes/dae-gc-refine-return.wast [hunk 4]
```diff
@@ -2,6 +2,12 @@
 ;; RUN: wasm-opt %s -all --dae -S -o - | filecheck %s
 ;; RUN: wasm-opt %s -all --dae --nominal -S -o - | filecheck %s --check-prefix NOMNL
 
+;; Note that the nulls in this module are wrapped in blocks, as without such
+;; wrapping they might be modified by the optimization (which will change the
+;; types of nulls in order to improve the LUB calculation). A block around the
+;; null stops that because we just look for a null (as we assume other opts
+;; would have remove anything unneeded before us).
```
[the remark this answers is not in the document]
Comment (kripken) [reply test/lit/passes/dae-gc-refine-return.wast]: Sounds good, done.

File: test/lit/passes/dae-gc-refine-params.wast (hunk 3 above)
[the remark this answers is not in the document]
Comment (tlively) [reply test/lit/passes/dae-gc-refine-params.wast:113]: Seems reasonable.

**Review by tlively — APPROVED** [review]

**Pull request merged**